indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -0,0 +1,399 @@
+      TaskToolbox toolbox,
+      String dataSource,
+      Interval interval
+  ) throws IOException, SegmentLoadingException
+  {
+    final List<DataSegment> usedSegments = toolbox.getTaskActionClient()
+                                                  .submit(new SegmentListUsedAction(dataSource, interval, null));
+    final Map<DataSegment, File> segmentFileMap = toolbox.fetchSegments(usedSegments);
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = VersionedIntervalTimeline
+        .forSegments(usedSegments)
+        .lookup(interval);
+    return Pair.of(segmentFileMap, timelineSegments);
+  }
+
+  private static DataSchema createDataSchema(
+      String dataSource,
+      Interval interval,
+      IndexIO indexIO,
+      ObjectMapper jsonMapper,
+      List<TimelineObjectHolder<String, DataSegment>> timelineSegments,
+      Map<DataSegment, File> segmentFileMap
+  )
+      throws IOException, SegmentLoadingException
+  {
+    // find metadata for interval
+    final List<QueryableIndex> queryableIndices = loadSegments(timelineSegments, segmentFileMap, indexIO);
+
+    // find merged aggregators
+    final List<AggregatorFactory[]> aggregatorFactories = queryableIndices
+        .stream()
+        .map(index -> index.getMetadata().getAggregators())
+        .collect(Collectors.toList());
+    final AggregatorFactory[] mergedAggregators = AggregatorFactory.mergeAggregators(aggregatorFactories);
+
+    if (mergedAggregators == null) {
+      throw new ISE("Failed to merge aggregators[%s]", aggregatorFactories);
+    }
+
+    // find granularity spec
+    // set rollup only if rollup is set for all segments
+    final boolean rollup = queryableIndices.stream().allMatch(index -> index.getMetadata().isRollup());
+    final GranularitySpec granularitySpec = new ArbitraryGranularitySpec(
+        new NoneGranularity(),
+        rollup,
+        ImmutableList.of(interval)
+    );
+
+    // find unique dimensions
+    final DimensionsSpec dimensionsSpec = createDimensionsSpec(queryableIndices);
+    final InputRowParser parser = new NoopInputRowParser(new TimeAndDimsParseSpec(null, dimensionsSpec));
+
+    return new DataSchema(
+        dataSource,
+        jsonMapper.convertValue(parser, JacksonUtils.TYPE_REFERENCE_MAP_STRING_OBJECT),
+        mergedAggregators,
+        granularitySpec,
+        jsonMapper
+    );
+  }
+
+  private static DimensionsSpec createDimensionsSpec(List<QueryableIndex> queryableIndices)
+  {
+    final BiMap<String, Integer> uniqueDims = HashBiMap.create();
+    final Map<String, DimensionSchema> dimensionSchemaMap = new HashMap<>();
+
+    // Here, we try to retain the order of dimensions as they were specified since the order of dimensions may be
+    // optimized for performance.
+    // Dimensions are extracted from the recent segments to olders because recent segments are likely to be queried more
```

Comment:
I think this would also have the effect giving recent segments precedence in terms of what type each dimension has (like if an older segment stored a dimension as String but newer ones store it as Long). Can you mention the ordering and type precedence in the docs somewhere?

Reply:
One question that leads to, would it be useful to allow users some control over ordering and types?

I'm thinking that maybe that overcomplicates the Compaction Task which is meant to be simple to express, and users can manually write a full batch task if they have column ordering/type requirements for the final compacted segments, but would like to see what your thoughts are on that.

Reply:
> I think this would also have the effect giving recent segments precedence in terms of what type each dimension has 

Good point. Added doc.

> One question that leads to, would it be useful to allow users some control over ordering and types?

I think it's a good idea. Users can write a full batch spec for full control, but someone might want to have little more control with compaction task because it's simple. I added this feature and a unit test for it.